### PR TITLE
Adding parameters for convergence tests

### PR DESCRIPTION
**Pull request by sebglane**

### CODE REVIEW

File: include/rotatingMHD/run_time_parameters.h [hunk 1]
```diff
@@ -542,81 +517,6 @@ template<typename Stream>
...
-
-/*!
- * @brief Method forwarding parameters to a stream object.
- *
- * @details This method does not add a `std::endl` to the stream at the end.
- */
-template<typename Stream>
-Stream& operator<<(Stream &stream, const ConvergenceTestParameters &prm);
-
-
```
Comment (sebglane) [inline include/rotatingMHD/run_time_parameters.h:619]: Deleting duplicate

File: source/run_time_parameters.cc [hunk 2]
```diff
@@ -299,134 +299,6 @@ Stream& operator<<(Stream &stream, const OutputControlParameters &prm)
...
-  internal::add_line(stream,
-                     "Time-step reduction factor",
-                     prm.timestep_reduction_factor);
-
-  internal::add_header(stream);
-
-  return (stream);
-}
-
-
```
Comment (sebglane) [inline source/run_time_parameters.cc:429]: Deleting duplicate

File: source/run_time_parameters.cc [hunk 3]
```diff
@@ -2363,11 +2235,6 @@ template std::ostream & RMHD::RunTimeParameters::operator<<
 template dealii::ConditionalOStream  & RMHD::RunTimeParameters::operator<<
 (dealii::ConditionalOStream &, const RMHD::RunTimeParameters::OutputControlParameters &);
 
-template std::ostream & RMHD::RunTimeParameters::operator<<
-(std::ostream &, const RMHD::RunTimeParameters::ConvergenceTestParameters &);
-template dealii::ConditionalOStream  & RMHD::RunTimeParameters::operator<<
-(dealii::ConditionalOStream &, const RMHD::RunTimeParameters::ConvergenceTestParameters &);
```
Comment (sebglane) [inline source/run_time_parameters.cc:2369]: Deleting duplicate

File: include/rotatingMHD/run_time_parameters.h [hunk 4]
```diff
@@ -185,32 +186,6 @@ enum class ProblemType
...
-  /*!
-   * @brief Temporal convergence test.
-   * @details Test to study the temporal discretization dependence of
-   * convergence for a given problem.
-   * @note Temporal convergence tests should be performed with a fine
-   * spatial discretization, *i. e.*, a triangulation with small enough cells.
-   */
-  temporal
-};
```
Comment (sebglane) [inline include/rotatingMHD/run_time_parameters.h:210]: Deleting duplicate

File: applications/GuermondNeumannBC.cc [hunk 5]
```diff
@@ -495,10 +496,10 @@ void Guermond<dim>::run()
...
   std::ostringstream tablefilename;
-  tablefilename << ((this->prm.convergence_test_parameters.convergence_test_type ==
-                      RunTimeParameters::ConvergenceTestType::spatial)
+  tablefilename << ((this->prm.convergence_test_parameters.test_type ==
+  									 ConvergenceTest::ConvergenceTestType::spatial)
                      ? "GuermondNeumannBC_SpatialTest"
-                     : ("GuermondNeumannBC_TemporalTest_Level" + std::to_string(this->prm.convergence_test_parameters.n_global_initial_refinements)))
+                     : ("GuermondNeumannBC_TemporalTest_Level" + std::to_string(this->prm.spatial_discretization_parameters.n_initial_global_refinements)))
                 << "_Re"
                 << this->prm.Re;
```
Comment (sebglane) [inline applications/GuermondNeumannBC.cc:505]: This applications compiles but does not execute because of #103

**Comment by sebglane** [comment]
Ready to go. I checked all convergence test except for `GuermondNeumannBC.cc` see comment above.

File: include/rotatingMHD/convergence_struct.h [hunk 6]
```diff
@@ -46,6 +46,109 @@ template<typename Stream, int dim>
...
+  /*!
+   * @brief Temporal convergence test.
+   * @details Test to study the temporal discretization dependence of
+   * convergence for a given problem.
+   * @note Temporal convergence tests should be performed with a fine
+   * spatial discretization, *i. e.*, a triangulation with small enough cells.
+   */
+  temporal = 0x0002,
+
+  spatio_temporal = spatial|temporal
```
Comment (j507) [inline include/rotatingMHD/convergence_struct.h:75]: Typo?

File: applications/Couette.prm [hunk 7]
```diff
@@ -103,6 +103,13 @@ subsection Output control parameters
 end
 
 
+subsection Refinement control parameters
```
Comment (j507) [inline applications/Couette.prm:106]: This subsection could use a renaming. Maybe  "Spatial discretization parameters"?

File: applications/Couette.prm [hunk 8]
```diff
@@ -103,6 +103,13 @@ subsection Output control parameters
...
+  set Adaptive mesh refinement               = false
+  set Maximum number of levels               = 10
+  set Number of initial global refinements   = 3
+end
+
+
 subsection Time stepping parameters
```
Comment (j507) [inline applications/Couette.prm:113]: This one too. Maybe "Temporal discretization parameters"?

File: source/convergence_struct.cc [hunk 9]
```diff
@@ -202,6 +202,214 @@ void ConvergenceAnalysisData<dim>::write_text(std::string filename) const
...
+}
+
+
+void ConvergenceTestParameters::declare_parameters(ParameterHandler &prm)
+{
+  prm.enter_subsection("Convergence test parameters");
+  {
+    prm.declare_entry("Convergence test type",
+                      "temporal",
+                      Patterns::Selection("spatial|temporal|both"));
```
Comment (j507) [inline source/convergence_struct.cc:300]: `both` is defined here but is not implemented in the code.

**Review by j507 — APPROVED** [review]

File: include/rotatingMHD/convergence_struct.h (hunk 6 above)
Comment (sebglane) [reply include/rotatingMHD/convergence_struct.h:75]: No

File: source/convergence_struct.cc (hunk 9 above)
Comment (sebglane) [reply source/convergence_struct.cc:300]: > `both` is defined here but is not implemented in the code.

This option is catched here

https://github.com/j507/RotatingMHD/blob/6f3db9e1c469afc24b5f2c18165f7258e5c17c12/source/convergence_struct.cc#L345-L362

In this PR, the new parameter structure is added but the way the convergence tests are performed is still *old*.

Comment (sebglane) [reply source/convergence_struct.cc:300]: In order to capture the spatio-temporal case, I added an assert in 36145cb.

**Pull request merged**